src/include/sof/audio/buffer.h

```diff
@@ -271,6 +271,48 @@ static inline uint32_t buffer_avail_frames(struct comp_buffer *source,
 	return MIN(src_frames, sink_frames);
 }
 
+/**
+ * Returns the contiguous memory regions which are free and available
+ * in this buffer.
+ * @param buf Component buffer
+ * @param[out] r_ptr Beginning of readable memory region
+ * @param[out] r_size Size (in bytes) of readable memory region
+ * @param[out] w_ptr Beginning of writeable memory region
+ * @param[out] w_size Size (in bytes) of writeable memory region
+ * @return 0 on success, -EINVAL if buf is NULL or all 4 output parameters are
+ *         NULL
+ */
+static inline int buffer_contig_avail_free(struct comp_buffer *buf,
+	void **r_ptr, size_t *r_size, void **w_ptr, size_t *w_size)
+{
+	char *read;
+	char *write;
+	size_t read_size;
+	size_t write_size;
+
+	if (!buf)
+		return -EINVAL;
+	if (!r_ptr && !r_size && !w_ptr && !w_size)
+		return -EINVAL;
+	read = buf->r_ptr;
+	read_size = buf->avail;
+	if (read + read_size > (char *)buf->end_addr)
+		read_size = (char *)buf->end_addr - read;
+	write = buf->w_ptr;
+	write_size = buf->free;
+	if (write + write_size > (char *)buf->end_addr)
+		write_size = (char *)buf->end_addr - write;
+	if (r_ptr)
+		*r_ptr = read;
+	if (r_size)
+		*r_size = read_size;
+	if (w_ptr)
+		*w_ptr = write;
+	if (w_size)
+		*w_size = write_size;
+	return 0;
+}
+
```

Comment:
This starts to be quite long as an inlined function. 

How about the combination of read and write. Would separate functions to query contig space for read and write be more suitable?

Reply:
Agree with @kv2019i - is there any specific reason why this should be an inline function? Do we really expect to use this function heavily on performance-critical paths? As for read / write separation - it would be easier to tell if we had this patch in the same series as any callers of this function.

Reply:
I think we could do a function for read and one for write as I see essentially no use case where I'd pass all 4 parameters as non-null.

Unfortunately I don't think I'd have any public interface soon enough (probably a rewritten variant will show up in March or something). Will separate them.

Reply:
Done the separation.

The function will be called up to 4 times per copy() call of a component (until either the source buffer is truly empty or the sink buffer is truly full). The processing function expects contiguous input and output buffers, that's why I made this pull request in the first place.

Unfortunately the processing function itself is not open source (hence I will need to focus on loadable modules again at some point, once I get a good design and the time on my hands) and I thus cannot update it to be able to use the actual circular buffers of SOF.

Some sample code within the `copy` function of that future component:

```c
struct comp_buffer *src_buf, *sink_buf;
void *src_addr, *sink_addr;
size_t src_size, sink_size;
size_t src_size2, sink_size2;
int ret;

do {
    ret = buffer_contig_avail(src_buf, &src_addr, &src_size);
    assert(!ret);
    if (!src_size)
        break;
    src_size2 = src_size;
    ret = buffer_contig_free(sink_buf, &sink_addr, &sink_size);
    assert(!ret);
    if (!sink_size)
        break;
    sink_size2 = sink_size;
    cd->Decode(src_addr, &src_size, sink_addr, &sink_size);
    comp_update_buffer_consume(src_buf, src_size2 - src_size);
    comp_update_buffer_produce(sink_buf, sink_size2 - sink_size);
} while (src_size && sink_size);
```

Reply:
@paulstelian97 thanks for this great work! Lets keep this commit in mind and merge it once it has a real user.

Reply:
@mrajwa @keyonjie doesn't KWD need continuous buffers ?
@dbaluta I've no problem merging now if you guys need this for some WiP feature.

Reply:
I am heavily against merging code that doesn't have any user, we plan to send patches using it in the next months. So, lets close this for now and merge it when it actually have a real user.

Reply:
@lgirdwood I am against on pushing code that doesn't immediately have a real user. I think we will get back on this in February. Lets close it for now.

@paulstelian97 thanks for posting the patch lets keep it in mind and added in the PR introducing the compressed playback/etc.

Reply:
ok, I'm fine as long as you guys are tracking this as API users. Did not want this to fall between the cracks.